### DISCUSSION TITLE
Theta hat doesn't have y offset

### DISCUSSION THREAD

**Discussion by mitchellh** [harfbuzz/harfbuzz#4525] [General] [answered]
Character: `θ̂`
Font: https://www.jetbrains.com/lp/mono/ (non-variable version)
Backends: Freetype and CoreText (both appear the same)
Harfbuzz version: 8.2.2, 8.3.0, and 74e897117196629f2f736b8053386cbf2e454e11 all exhibit the issue

```
$ hb-shape --output-format=json ~/Library/Fonts/JetBrainsMono-Regular.ttf θ̂ | jq .
[
  {
    "g": "theta",
    "cl": 0,
    "dx": 0,
    "dy": 0,
    "ax": 600,
    "ay": 0
  },
  {
    "g": "uni0302",
    "cl": 0,
    "dx": 0,
    "dy": 0,
    "ax": 0,
    "ay": 0
  }
]
```

If I follow the advance directly and render, the rendering appears like this (enlarged for effect):

![CleanShot 2023-12-11 at 20 58 08@2x](https://github.com/harfbuzz/harfbuzz/assets/1299/1c17f0e6-6fd6-4648-86e9-e2650c5810bc)

Notice the hat is inside the theta. I reimplemented my shaping using CoreText directly for shaping, and it provides the following information:

```
codepoint=952 glyph=445 pos=graphics.geometry.Point{ .x = 0.0e+00, .y = 0.0e+00 } advance=graphics.geometry.Size{ .width = 1.8e+01, .height = 5.2875000000000005e+00 } index=0 cluster=0
codepoint=770 glyph=372 pos=graphics.geometry.Point{ .x = 1.8e+01, .y = 5.2875000000000005e+00 } advance=graphics.geometry.Size{ .width = 0.0e+00, .height = -5.2875000000000005e+00 } index=1 cluster=0
```

Note the y advance being set (`advance.height`) and then being unset. This results in the theta-hat rendering correctly. Other applications (unsure if they're using hb) appear to render theta-hat correctly with this font as well. I'm unsure if I'm just configuring Harfbuzz incorrectly, but I'm unable to find a solution to this. Thank you for an excellent library.

**Comment by khaledhosny** [comment, marked as answer]
This is a font bug. The font does not have mark positioning for the `theta` glyph, but it has it for other glyphs. Core Text seems to be doing some fallback mark positioning here, but HarfBuzz does not do this unless the font has no mark positioning data at all.

**Reply by mitchellh** [reply]
Got it, I'll report the issue to JB. Thank you.

EDIT: Reported here: https://github.com/JetBrains/JetBrainsMono/issues/660

**Comment by behdad** [comment]
> This is a font bug. The font does not have mark positioning for the `theta` glyph, but it has it for other glyphs. Core Text seems to be doing some fallback mark positioning here, but HarfBuzz does not do this unless the font has no mark positioning data at all.

It's really curious that CoreText does this level of fallback, since the glyph might be designed to just fit... Maybe @nedley can tell us what CoreText is doing.

**Reply by mitchellh** [reply]
If it's helpful, I did some additional testing. The following fonts all render θ̂ correctly with CoreText shaping and **do not with HarfBuzz**. I tried to pick fonts that were built-in to the system. 

- Andale Mono
- Menlo
- Monaco
- Courier New

It feels strange to assume all of these fonts have bugs...?

**Reply by khaledhosny** [reply]
* Andale Mono: does not have combining circumflex
* Menlo: looks fine:
  ![menlo](https://github.com/harfbuzz/harfbuzz/assets/93914/ee09cd09-3641-4469-bb4b-20c64520dcc1)
* Monoca: Does not have theta
* Courier New: Has mark positioning only for Arabic and Hebrew, and additionally claasifies combining circumflex as base glyph.

**Reply by nedley** [reply]
CoreText may perform fallback mark positioning when a mark glyph has not been substituted or attached.

**Reply by behdad** [reply]
Humm. Interesting. Thanks.

I know in the past the argument against doing that for us was "what if the designer put it in the right place by design?". I'll experiment a bit and see if we can do the same without breaking the world.  AFAIU DWrite doesn't do that.

We probably decide this on a per-shaper basis...

cc @dscorbett

**Reply by jfkthame** [reply]
As a possible heuristic: maybe try this *only* for marks with CCC=220 (below) or CCC=230 (above), not other classes; and perhaps initially just in the generic shaper, not in the script-specific ones.

**Comment by behdad** [comment]
What system and what HB version are you testing on?

> * Andale Mono
> * Menlo

With Menlo on MacOS 13.3.1 (22E261) it works fine with HarfBuzz for me. I don't get the exact same positioning as CoreText, but it doesn't *look* wrong.

> * Monaco

My Mac's Monaco doesn't support theta.

> * Courier New
> 
> It feels strange to assume all of these fonts have bugs...?
